File: trunk/Client/Assets/Scripts/Common/Rpc/COM_BattleRoom.cs
public class COM_BattleRoom{
  public virtual void Serialize(IWriter w){
    Mask mask = new Mask(1);
    mask.WriteBit(InstId!=0);
    mask.WriteBit(Status!=0);
    mask.WriteBit(PlayerList!=null&&PlayerList.Length!=0);
    mask.WriteBit(true); // Target
    mask.WriteBit(Bout!=0);
    mask.WriteBit(TurnMove!=0);
    mask.WriteBit(true); // NextPlayer
    w.Write(mask.Bytes);
    //S InstId
    if(InstId!=0){
      w.Write(InstId);
    }
    //S Status
    if(Status!=0){
      w.Write(Status);
    }
    //S PlayerList
    if(PlayerList!=null&&PlayerList.Length!=0){
      w.WriteSize(PlayerList.Length);
      for(int i=0; i<PlayerList.Length; ++i){
        PlayerList[i].Serialize(w);
      }
    }
    //S Target
    {
      Target.Serialize(w);
    }
    //S Bout
    if(Bout!=0){
      w.Write(Bout);
    }
    //S TurnMove
    if(TurnMove!=0){
      w.Write(TurnMove);
    }
    //S NextPlayer
    {
      NextPlayer.Serialize(w);
    }
  }
  public virtual bool Deserialize(IReader r){
    Mask mask = new Mask(1);
    if(!r.Read(ref mask.Bytes)){
      return false;
    }
    //D InstId
    if(mask.ReadBit()){
      if(!r.Read(ref InstId)){
        return false;
      }
    }
    //D Status
    if(mask.ReadBit()){
      if(!r.Read(ref Status)){
        return false;
      }
    }
    //D PlayerList
    if(mask.ReadBit()){
      int size = 0;
      if(!r.ReadSize(ref size) || size > 255){
        return false;
      }
      PlayerList = new COM_BattlePlayer[size];
      for(int i=0; i<size; ++i){
        PlayerList[i] = new COM_BattlePlayer();
        if(!PlayerList[i].Deserialize(r)){
          return false;
        }
      }
    }
    //D Target
    if(mask.ReadBit()){
      if(Target==null){
        Target = new COM_BattlePlayer();
      }
      if(!Target.Deserialize(r)){
        return false;
      }
    }
    //D Bout
    if(mask.ReadBit()){
      if(!r.Read(ref Bout)){
        return false;
      }
    }
    //D TurnMove
    if(mask.ReadBit()){
      if(!r.Read(ref TurnMove)){
        return false;
      }
    }
    //D NextPlayer
    if(mask.ReadBit()){
      if(NextPlayer==null){
        NextPlayer = new COM_BattlePlayer();
      }
      if(!NextPlayer.Deserialize(r)){
        return false;
      }
    }
    return true;
  }
  public long InstId;
  public int Status;
  public COM_BattlePlayer[] PlayerList;
  public COM_BattlePlayer Target;
  public int Bout;
  public int TurnMove;
  public COM_BattlePlayer NextPlayer;
}

File: trunk/server/src/logic/prpc/COM_BattleRoom.go
package prpc

import (
	"bytes"
	"encoding/json"
	"sync"
)

type COM_BattleRoom struct {
	sync.Mutex
	InstId     int64              //0
	Status     int32              //1
	PlayerList []COM_BattlePlayer //2
	Target     COM_BattlePlayer   //3
	Bout       int32              //4
	TurnMove   int32              //5
	NextPlayer COM_BattlePlayer   //6
}

func (this *COM_BattleRoom) SetInstId(value int64) {
	this.Lock()
	defer this.Unlock()
	this.InstId = value
}
func (this *COM_BattleRoom) GetInstId() int64 {
	this.Lock()
	defer this.Unlock()
	return this.InstId
}
func (this *COM_BattleRoom) SetStatus(value int32) {
	this.Lock()
	defer this.Unlock()
	this.Status = value
}
func (this *COM_BattleRoom) GetStatus() int32 {
	this.Lock()
	defer this.Unlock()
	return this.Status
}
func (this *COM_BattleRoom) SetPlayerList(value []COM_BattlePlayer) {
	this.Lock()
	defer this.Unlock()
	this.PlayerList = value
}
func (this *COM_BattleRoom) GetPlayerList() []COM_BattlePlayer {
	this.Lock()
	defer this.Unlock()
	return this.PlayerList
}
func (this *COM_BattleRoom) SetTarget(value COM_BattlePlayer) {
	this.Lock()
	defer this.Unlock()
	this.Target = value
}
func (this *COM_BattleRoom) GetTarget() COM_BattlePlayer {
	this.Lock()
	defer this.Unlock()
	return this.Target
}
func (this *COM_BattleRoom) SetBout(value int32) {
	this.Lock()
	defer this.Unlock()
	this.Bout = value
}
func (this *COM_BattleRoom) GetBout() int32 {
	this.Lock()
	defer this.Unlock()
	return this.Bout
}
func (this *COM_BattleRoom) SetTurnMove(value int32) {
	this.Lock()
	defer this.Unlock()
	this.TurnMove = value
}
func (this *COM_BattleRoom) GetTurnMove() int32 {
	this.Lock()
	defer this.Unlock()
	return this.TurnMove
}
func (this *COM_BattleRoom) SetNextPlayer(value COM_BattlePlayer) {
	this.Lock()
	defer this.Unlock()
	this.NextPlayer = value
}
func (this *COM_BattleRoom) GetNextPlayer() COM_BattlePlayer {
	this.Lock()
	defer this.Unlock()
	return this.NextPlayer
}
func (this *COM_BattleRoom) Serialize(buffer *bytes.Buffer) error {
	this.Lock()
	defer this.Unlock()
	//field mask
	mask := newMask1(1)
	mask.writeBit(this.InstId != 0)
	mask.writeBit(this.Status != 0)
	mask.writeBit(len(this.PlayerList) != 0)
	mask.writeBit(true) //Target
	mask.writeBit(this.Bout != 0)
	mask.writeBit(this.TurnMove != 0)
	mask.writeBit(true) //NextPlayer
	{
		err := write(buffer, mask.bytes())
		if err != nil {
			return err
		}
	}
	// serialize InstId
	{
		if this.InstId != 0 {
			err := write(buffer, this.InstId)
			if err != nil {
				return err
			}
		}
	}
	// serialize Status
	{
		if this.Status != 0 {
			err := write(buffer, this.Status)
			if err != nil {
				return err
			}
		}
	}
	// serialize PlayerList
	if len(this.PlayerList) != 0 {
		{
			err := write(buffer, uint(len(this.PlayerList)))
			if err != nil {
				return err
			}
		}
		for _, value := range this.PlayerList {
			err := value.Serialize(buffer)
			if err != nil {
				return err
			}
		}
	}
	// serialize Target
	{
		err := this.Target.Serialize(buffer)
		if err != nil {
			return err
		}
	}
	// serialize Bout
	{
		if this.Bout != 0 {
			err := write(buffer, this.Bout)
			if err != nil {
				return err
			}
		}
	}
	// serialize TurnMove
	{
		if this.TurnMove != 0 {
			err := write(buffer, this.TurnMove)
			if err != nil {
				return err
			}
		}
	}
	// serialize NextPlayer
	{
		err := this.NextPlayer.Serialize(buffer)
		if err != nil {
			return err
		}
	}
	return nil
}
func (this *COM_BattleRoom) Deserialize(buffer *bytes.Buffer) error {
	this.Lock()
	defer this.Unlock()
	//field mask
	mask, err := newMask0(buffer, 1)
	if err != nil {
		return err
	}
	// deserialize InstId
	if mask.readBit() {
		err := read(buffer, &this.InstId)
		if err != nil {
			return err
		}
	}
	// deserialize Status
	if mask.readBit() {
		err := read(buffer, &this.Status)
		if err != nil {
			return err
		}
	}
	// deserialize PlayerList
	if mask.readBit() {
		var size uint
		err := read(buffer, &size)
		if err != nil {
			return err
		}
		this.PlayerList = make([]COM_BattlePlayer, size)
		for i, _ := range this.PlayerList {
			err := this.PlayerList[i].Deserialize(buffer)
			if err != nil {
				return err
			}
		}
	}
	// deserialize Target
	if mask.readBit() {
		err := this.Target.Deserialize(buffer)
		if err != nil {
			return err
		}
	}
	// deserialize Bout
	if mask.readBit() {
		err := read(buffer, &this.Bout)
		if err != nil {
			return err
		}
	}
	// deserialize TurnMove
	if mask.readBit() {
		err := read(buffer, &this.TurnMove)
		if err != nil {
			return err
		}
	}
	// deserialize NextPlayer
	if mask.readBit() {
		err := this.NextPlayer.Deserialize(buffer)
		if err != nil {
			return err
		}
	}
	return nil
}
func (this *COM_BattleRoom) String() string {
	b, e := json.Marshal(this)
	if e != nil {
		return e.Error()
	} else {
		return string(b)
	}
}
